components/script/dom/notification.rs

```diff
@@ -0,0 +1,444 @@
+            // A non-persistent notification is a notification whose service worker registration is null.
+            serviceworker_registration: None,
+            title,
+            body,
+            data,
+            dir,
+            image,
+            icon,
+            badge,
+            lang,
+            silent,
+            origin,
+            vibration_pattern,
+            timestamp,
+            renotify,
+            tag,
+            require_interaction,
+            actions,
+        })
+    }
+}
+
+impl NotificationMethods<crate::DomTypeHolder> for Notification {
+    /// <https://notifications.spec.whatwg.org/#constructors>
+    fn Constructor(
+        global: &GlobalScope,
+        proto: Option<HandleObject>,
+        can_gc: CanGc,
+        title: DOMString,
+        options: RootedTraceableBox<NotificationOptions>,
+    ) -> Fallible<DomRoot<Notification>> {
+        // step 1: Check global is a ServiceWorkerGlobalScope
+        if global.downcast::<ServiceWorkerGlobalScope>().is_some() {
+            return Err(Error::Type(
+                "Notification constructor cannot be used in ServiceWorkerGlobalScope.".to_string(),
+            ));
+        }
+
+        // step 2: Check options.actions must be empty
+        if !options.actions.is_empty() {
+            return Err(Error::Type(
+                "Actions are only supported for persistent notifications.".to_string(),
+            ));
+        }
+
+        // step 3: Create a notification with a settings object
+        // https://notifications.spec.whatwg.org/#create-a-notification-with-a-settings-object
+        let fallback_timestamp = global.performance().Now().to_u64().unwrap();
+        let notification = Notification::new(
+            global,
+            proto,
+            can_gc,
+            title,
+            options,
+            None,
+            None,
+            fallback_timestamp,
+        )?;
+
+        // FIXME: Run step 5.1, 5.2 in parallel
+        // step 5.1: If the result of getting the notifications permission state is not "granted",
+        //           then queue a task to fire an event named error on this, and abort these steps.
+        // TODO: `get_descriptor_permission_state` seems always assume
+        //       we are in non-secure environment and return "Prompt"
+        let permission_state =
+            get_descriptor_permission_state(PermissionName::Notifications, Some(global));
+        if permission_state != PermissionState::Granted {
+            notification
+                .upcast::<EventTarget>()
+                .fire_event(atom!("error"), CanGc::note());
+            // TODO: abort steps
+        }
+        // TODO: step 5.2: Run the notification show steps for notification
+        // https://notifications.spec.whatwg.org/#notification-show-steps
+
+        Ok(notification)
+    }
+
+    /// <https://notifications.spec.whatwg.org/#get-the-notifications-permission-state>
```

Comment:
The right link for this is `https://notifications.spec.whatwg.org/#dom-notification-permission`, and it would be best to put the logic of #get-the-notifications-permission-state in separate method, because it is used in different places in the spec. 

Also here below you are doing the second step, but you're skipping it in the constructor above, so if you were to put the logic in a method and use it in both places you'd be more spec compliant. 

Note that this matters because "default" could be set to "granted", see https://w3c.github.io/permissions/#dfn-default-state